lib/Transforms/YosysOptimizer/RTLILImporter.cpp

```diff
@@ -0,0 +1,162 @@
+  llvm::MapVector<Wire *, SmallVector<Value>> retBitValues;
+  for (auto *wire : module->wires()) {
+    // The RTLIL module may also have intermediate wires that are neither inputs
+    // nor outputs.
+    if (wire->port_input) {
+      argTypes.push_back(builder.getIntegerType(wire->width));
+      wireArgs.push_back(wire);
+    } else if (wire->port_output) {
+      retTypes.push_back(builder.getIntegerType(wire->width));
+      wireRet.push_back(wire);
+      retBitValues[wire].resize(wire->width);
+    }
+  }
+
+  // Build function
+  FunctionType funcType = builder.getFunctionType(argTypes, retTypes);
+  auto function = func::FuncOp::create(
+      builder.getUnknownLoc(), module->name.str().replace(0, 1, ""), funcType);
+
+  auto *block = function.addEntryBlock();
+  auto b = ImplicitLocOpBuilder::atBlockBegin(function.getLoc(), block);
+
+  // Map the RTLIL wires to the block arguments' Values.
+  for (auto i = 0; i < wireArgs.size(); i++) {
+    addWireValue(wireArgs[i], block->getArgument(i));
+  }
+
+  // Convert cells to Operations according to topological order.
+  for (const auto &cellName : cellOrdering) {
+    assert(module->cells_.count("\\" + cellName) != 0 &&
+           "expected cell in RTLIL design");
+    auto *cell = module->cells_["\\" + cellName];
+
+    SmallVector<Value, 4> inputValues;
+    for (const auto &conn : getInputs(cell)) {
+      inputValues.push_back(getBit(conn, b));
+    }
+    auto *op = createOp(cell, inputValues, b);
+    auto resultConn = getOutput(cell);
+    addResultBit(resultConn, op->getResult(0), retBitValues);
+  }
+
+  // Wire up remaining connections.
+  for (const auto &conn : module->connections()) {
+    auto output = conn.first;
+    // These must be output wire connections (either an output bit or a bit of a
+    // multi-bit output wire).
+    assert(output.is_wire() || output.as_bit().is_wire());
+    assert(retBitValues.contains(output.as_wire()) ||
+           retBitValues.contains(output.as_bit().wire));
+    if (conn.second.chunks().size() == 1) {
+      Value connValue = getBit(conn.second, b);
+      addResultBit(output, connValue, retBitValues);
+    } else {
+      auto chunks = conn.second.chunks();
+      for (size_t i = 0; i < output.size(); i++) {
+        Value connValue = getBit(conn.second.bits().at(i), b);
+        addResultBit(output.bits().at(i), connValue, retBitValues);
+      }
+    }
+  }
+
+  // Concatenate result bits if needed, and return result.
+  SmallVector<Value, 4> returnValues;
+  for (const auto &[resultWire, retBits] : retBitValues) {
+    if (retBits.size() > 1) {
+      auto concatOp = b.create<comb::ConcatOp>(retBits);
+      returnValues.push_back(concatOp.getResult());
+    } else {
+      returnValues.push_back(getWireValue(resultWire));
+    }
+  }
+  b.create<func::ReturnOp>(returnValues);
+
+  // OperationFolder folder(context);
+  // folder.tryToFold(function);  // TODO: add failed(), LogicalResult
```

Comment:
@j2kun: QQ here: the resulting function has redundant constants as I created the arguments for each op - I expected this folder to de-dupe them, but this didn't end up working. Have you ever come across anything else? The builder also uses `b.createOrFold<arith::ConstantOp>` where necessary.

Reply:
Maybe I should be using this for constant insertion https://mlir.llvm.org/doxygen/classmlir_1_1OperationFolder.html#a3573a3dcbb3cbb1b3f48b57c358851c1 - I'll give it a shot